File: siege-engine/frontend/src/api/references.ts
import { z } from 'zod';
import api from './client';

// Project references layer (read-only on the dashboard). Refs are
// supplemental documents (DSL specs, deployment runbooks,
// cross-component invariants) whose body lives in the project's
// git repo at `refs/<ref_id>/body.md`. Any node that draws a
// `reference` edge at a ref pulls its content into its regen
// context.
//
// Authoring happens in Claude Code via the `/create_ref` skill;
// the dashboard reads the projected state.

export const ReferenceEdgeSchema = z.object({
  edge_id: z.string(),
  source_id: z.string(),
  target_id: z.string(),
});
export type ReferenceEdge = z.infer<typeof ReferenceEdgeSchema>;

export const ReferenceSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  has_content: z.boolean(),
  updated_at: z.string(),
});
export type ReferenceSummary = z.infer<typeof ReferenceSummarySchema>;

export const ReferenceListResponseSchema = z.object({
  references: z.array(ReferenceSummarySchema),
});
export type ReferenceListResponse = z.infer<typeof ReferenceListResponseSchema>;

const ReferenceDetailRawSchema = z.object({
  node: z.object({
    id: z.string(),
    name: z.string(),
    content: z.string(),
    updated_at: z.string(),
  }),
  outgoing_edges: z.array(ReferenceEdgeSchema),
  incoming_edges: z.array(ReferenceEdgeSchema),
});

export type ReferenceDetail = z.infer<typeof ReferenceDetailRawSchema>;

export function makeReferencesApi(projectId: string) {
  return {
    async list(): Promise<ReferenceListResponse> {
      const { data } = await api.get(`/projects/${projectId}/references`);
      return ReferenceListResponseSchema.parse(data);
    },

    async getDetail(refId: string): Promise<ReferenceDetail> {
      const { data } = await api.get(`/projects/${projectId}/references/${refId}`);
      return ReferenceDetailRawSchema.parse(data);
    },
  };
}

// Parse a stored `<reference>` XML block into structured fields
// for display. Server-validated content; client-side extraction
// is regex-based to match the minimal parser used by vocab.
export interface ParsedReference {
  title: string;
  body: string;
  seeAlsoIds: string[];
}

export function parseReference(content: string): ParsedReference {
  const title = extractTagText(content, 'title') ?? '';
  const body = extractTagText(content, 'body') ?? '';
  const seeAlsoIds = extractSeeAlsoIds(content);
  return { title, body, seeAlsoIds };
}

function extractTagText(xml: string, tag: string): string | null {
  const match = xml.match(new RegExp(`<${tag}>([\\s\\S]*?)</${tag}>`));
  if (!match) return null;
  return unescapeXml(match[1].trim());
}

function extractSeeAlsoIds(xml: string): string[] {
  const block = xml.match(/<see-also>([\s\S]*?)<\/see-also>/);
  if (!block) return [];
  const ids: string[] = [];
  const refRe = /<ref\s+to="([^"]*)"/g;
  let m: RegExpExecArray | null;
  while ((m = refRe.exec(block[1])) !== null) {
    ids.push(m[1]);
  }
  return ids;
}

function unescapeXml(text: string): string {
  return text.replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&amp;/g, '&');
}

File: siege-engine/backend/graph/references.py
"""Project reference node lookups.

Reference entries are ``ref_*`` nodes — a first-class node tier
carrying supplemental documents (DSL specs, deployment runbooks,
cross-component invariants) whose bodies live in the project repo
at ``refs/<ref_id>/body.md`` (v3 git-backed). The dashboard reads
the projected state via these helpers; authoring happens in
Claude Code via the ``/create_ref`` skill.

Unlike vocabulary (scoped via ``parent_id`` to project or feature),
refs are always top-level: the reducer
(``_enforce_reference_parent_constraint``) rejects any attempt to
parent a ref under another node.
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.orm import Session

from backend.models.node import Edge, Node

REF_TIER = "ref"
REFERENCE_EDGE_TYPE = "reference"


def list_project_references(session: Session, project_id: str) -> list[Node]:
    """Return every ``ref_*`` node in the project, ordered by name."""
    return list(
        session.execute(
            select(Node)
            .where(
                Node.project_id == project_id,
                Node.tier == REF_TIER,
            )
            .order_by(Node.name.asc())
        ).scalars()
    )


def reference_by_id(session: Session, ref_id: str) -> Node | None:
    """Return the ``ref_*`` node with ``id == ref_id``, or ``None``."""
    node = session.get(Node, ref_id)
    if node is None or node.tier != REF_TIER:
        return None
    return node


def reference_by_name(session: Session, project_id: str, name: str) -> Node | None:
    """Look up a ref by its title/name within the project."""
    return session.execute(
        select(Node).where(
            Node.project_id == project_id,
            Node.tier == REF_TIER,
            Node.name == name,
        )
    ).scalar_one_or_none()


def outgoing_reference_edges(session: Session, project_id: str, source_id: str) -> list[Edge]:
    """Return every ``reference`` edge whose source is ``source_id``."""
    return list(
        session.execute(
            select(Edge).where(
                Edge.project_id == project_id,
                Edge.edge_type == REFERENCE_EDGE_TYPE,
                Edge.source_id == source_id,
            )
        ).scalars()
    )


def incoming_reference_edges(session: Session, project_id: str, target_id: str) -> list[Edge]:
    """Return every ``reference`` edge whose target is ``target_id``."""
    return list(
        session.execute(
            select(Edge).where(
                Edge.project_id == project_id,
                Edge.edge_type == REFERENCE_EDGE_TYPE,
                Edge.target_id == target_id,
            )
        ).scalars()
    )
